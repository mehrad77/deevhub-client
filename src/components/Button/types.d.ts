import { TColors } from '../../types';

export interface IStyledButton {
  kind?: 'filled' | 'outlined' | 'dashed';
  color?: TColors;
  halfBorderRadius?: 'left' | 'right' | 'both';
  // TODO: remove these `heiht` and `width` when the StyleGuide was ready
  height?: string;
  width?: string;
  padding?: string;
}

export interface IButtonProps extends IStyledButton {
  /** Text value of the button */
  text: string;
  //
  type?: 'submit';
  //
  loading?: boolean;
  /** A function that will be used on `onClick` event */
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
  /** If there was no `onClick` prop, this 'href' value
   * will be used to redirect use to the given href
   **/
  href?: string;
  /** Button `disabled` prop **/
  disabled?: boolean;
  /** Pass extra style object **/
  style?: React.CSSProperties;
  /** spesify the classname of the Button **/
  className?: string;
}
