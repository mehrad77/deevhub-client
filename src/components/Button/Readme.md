# Button

```js
<Button text="Push Me"  />

<Button text="With diffrent color" color="green" />

<Button text="With diffrent color" color="green"  />
```
