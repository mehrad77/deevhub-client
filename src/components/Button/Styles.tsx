import styled, { css } from 'styled-components';
import { prop, switchProp, theme } from 'styled-tools';
import { IStyledButton } from './types';

export const StyledButton = styled.button<IStyledButton>`
  height: ${prop('height', '40px')};
  width: ${prop('width', 'auto')};
  padding: ${prop('padding', 'auto')};
  background: ${({ kind, color }) =>
    kind === 'filled'
      ? theme('colors.' + color)
      : kind === 'outlined'
      ? theme('colors.backgroundWhite')
      : theme('colors.backgroundWhite')};

  .ButtonText {
    color: ${({ kind, color }) =>
      kind === 'filled'
        ? theme('colors.backgroundWhite')
        : kind === 'outlined'
        ? theme('colors.' + color)
        : theme('colors.textWhite')};
  }
  border: ${({ kind, color }) =>
    kind === 'filled'
      ? 'none'
      : kind === 'outlined'
      ? theme('border.' + color)
      : kind === 'dashed'
      ? theme('border.' + color + 'Dashed')
      : 'none'};
  border-radius: ${({ theme }) => theme.borderRadius};
  ${switchProp('halfBorderRadius', {
    left: css`
      border-top-left-radius: 0px;
      border-bottom-left-radius: 0px;
    `,
    right: css`
      border-top-right-radius: 0px;
      border-bottom-right-radius: 0px;
    `,
    both: css`
      border-radius: 0px;
    `,
  })}

  /* TOOD: sort below styles  */
  cursor: pointer;
  transition: 0.05s all;
  :active {
    transform: scale(0.98);
  }

  /* A11Y styles */
  :focus {
    outline: 0;
    box-shadow: 0 0 1px 1px ${({ kind = 'filled', color }) => theme('colors.' + color)}55;
  }
  ::-moz-focus-inner {
    border: 0;
  }
`;
