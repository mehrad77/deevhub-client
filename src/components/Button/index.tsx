import React, { FC } from 'react';
import { Text, Icon } from '..';
import { StyledButton } from './Styles';
import { IButtonProps } from './types';

const Button: FC<IButtonProps> = ({
  text,
  color,
  loading,
  onClick,
  href = '',
  ...rest
}) => {
  return (
    <StyledButton {...rest} color={color ? color : 'blue'} onClick={onClick}>
      {!loading && <Text className="ButtonText">{text}</Text>}
      {loading && <Icon type="loading" />}
    </StyledButton>
  );
};

Button.defaultProps = {
  kind: 'filled',
  color: 'blue',
};

export default React.memo(Button);
