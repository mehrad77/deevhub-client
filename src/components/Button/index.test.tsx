import React from 'react';
import { shallow, render } from 'enzyme';
import Button from '.';

const getComponent = () => {
  return render(<Button text="With green color" color="green" />);
};

describe('Button', () => {
  it('Renders button with green color', () => {
    const button = getComponent();
    expect(button).toMatchSnapshot();
  });

  it('Renders outlined button with default color', () => {
    const button = render(<Button text="Skip Changes" kind="outlined" width="150px" />);
    expect(button).toMatchSnapshot();
  });
});
