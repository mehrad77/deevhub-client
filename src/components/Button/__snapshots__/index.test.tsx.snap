// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Renders button with green color 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-jTzLTM kwFwXW",
      "color": "green",
      "kind": "filled",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "sc-bxivhb bcCefa ButtonText",
        },
        "children": Array [
          Object {
            "data": "With green color",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "color": undefined,
      "kind": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "color": undefined,
      "kind": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Button Renders outlined button with default color 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-jTzLTM ddlFZH",
      "color": "blue",
      "kind": "outlined",
      "width": "150px",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "sc-bxivhb bcCefa ButtonText",
        },
        "children": Array [
          Object {
            "data": "Skip Changes",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "color": undefined,
      "kind": undefined,
      "width": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "color": undefined,
      "kind": undefined,
      "width": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
